datafusion/physical-plan/src/sorts/sort_preserving_merge.rs

```diff
@@ -98,19 +99,56 @@ pub struct SortPreservingMergeExec {
     ///
     /// See [`Self::with_round_robin_repartition`] for more information.
     enable_round_robin_repartition: bool,
+    ///
+    progressive_eval_exec: Option<Arc<ProgressiveEvalExec>>,
 }
 
 impl SortPreservingMergeExec {
     /// Create a new sort execution plan
     pub fn new(expr: LexOrdering, input: Arc<dyn ExecutionPlan>) -> Self {
         let cache = Self::compute_properties(&input, expr.clone());
+
+        // Todo: check if partition statistic is accurate
+        // Organize the input partitions into chains,
+        // where elements of each chain are input partitions that are
+        // non-overlapping, and each chain is ordered internally by their min/max statistics.
+        let partition_groups = input
+            .statistics_by_partition()
+            .and_then(|stats| {
+                let min_max =
+                    MinMaxStatistics::new_from_statistics(&expr, &input.schema(), &stats);
+                min_max
+            })
+            .map(|min_max_stats| {
+                let res = min_max_stats.first_fit();
+                res
+            })
+            .inspect_err(|e| {
+                log::debug!(
+                    "error analyzing statistics: {e}\n falling back to full sort-merge"
+                )
+            })
+            .ok()
+            .filter(|groups| {
+                groups.len() < input.properties().partitioning.partition_count()
+            });
+        let mut progressive_eval_exec = None;
+        if let Some(partition_groups) = partition_groups {
+            progressive_eval_exec = Some(Arc::new(ProgressiveEvalExec::new(
+                input.clone(),
+                None,
+                None,
+                partition_groups,
+            )));
+        }
         Self {
             input,
             expr,
             metrics: ExecutionPlanMetricsSet::new(),
             fetch: None,
             cache,
             enable_round_robin_repartition: true,
+            progressive_eval_exec,
```

Comment:
We are instead replacing the SPM node with a ProgressEval node during an optimizer run. I assume this is a WIP and that will occur eventually?

Reply:
Yes, doing it in Optimizer makes sense to me